File: src/weaponsprocurement/gui/StockReviewLaunchState.java
package weaponsprocurement.gui;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class StockReviewLaunchState {
    private final StockReviewState state;
    private final List<StockReviewPendingPurchase> pendingTrades;
    private final boolean reviewMode;

    StockReviewLaunchState(StockReviewState state,
                           List<StockReviewPendingPurchase> pendingTrades,
                           boolean reviewMode) {
        this.state = state == null ? null : new StockReviewState(state);
        this.pendingTrades = immutableCopy(pendingTrades);
        this.reviewMode = reviewMode;
    }

    StockReviewState getState() {
        return state;
    }

    List<StockReviewPendingPurchase> getPendingTrades() {
        return pendingTrades;
    }

    boolean isReviewMode() {
        return reviewMode;
    }

    private static List<StockReviewPendingPurchase> immutableCopy(List<StockReviewPendingPurchase> source) {
        if (source == null || source.isEmpty()) {
            return Collections.emptyList();
        }
        List<StockReviewPendingPurchase> result = new ArrayList<StockReviewPendingPurchase>();
        for (int i = 0; i < source.size(); i++) {
            StockReviewPendingPurchase purchase = source.get(i);
            if (purchase != null) {
                result.add(purchase.copy());
            }
        }
        return Collections.unmodifiableList(result);
    }
}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/state/StockReviewLaunchState.kt
package weaponsprocurement.ui.stockreview.state

import weaponsprocurement.ui.stockreview.trade.StockReviewPendingTrade
import java.util.ArrayList
import java.util.Collections

class StockReviewLaunchState(
    state: StockReviewState?,
    pendingTrades: List<StockReviewPendingTrade>?,
    private val reviewMode: Boolean,
) {
    private val state: StockReviewState? = if (state == null) null else StockReviewState(state)
    private val pendingTrades: List<StockReviewPendingTrade> = immutableCopy(pendingTrades)

    fun getState(): StockReviewState? = state

    fun getPendingTrades(): List<StockReviewPendingTrade> = pendingTrades

    fun isReviewMode(): Boolean = reviewMode

    companion object {
        private fun immutableCopy(source: List<StockReviewPendingTrade>?): List<StockReviewPendingTrade> {
            if (source.isNullOrEmpty()) {
                return emptyList()
            }
            val result = ArrayList<StockReviewPendingTrade>()
            for (trade in source) {
                val copy = trade?.copy()
                if (copy != null) {
                    result.add(copy)
                }
            }
            return Collections.unmodifiableList(result)
        }
    }
}